epitome/models.py

```diff
@@ -746,3 +750,149 @@ def create_model(self, **kwargs):
+        radii = self.radii
+
+        s = time.time()
+
+        stacked = np.stack([to_stack] * accessibility_peak_matrix.shape[0], axis=0)
+        names = stacked[:, :, 1]
+        to_stack = stacked[:, :, 0]
+        to_stack = np.expand_dims(to_stack, axis=-1)
+
+        same_size = accessibility_peak_matrix.shape[1] == len(conversionObject.joined.idx_base)
+
+        if not same_size:
+            added_indices = []
+            old_idx, counter, old_i = 0, 0, 0
+            indices_to_merge = []
+            for ctr, (i, i_base) in enumerate(zip(conversionObject.joined.idx, conversionObject.joined.idx_base)):
+                if i_base == -1:
+                    continue
+                if i != old_i:
+                    indices_to_merge.append((old_idx, counter))
+                    old_idx = counter
+                added_indices.append(accessibility_peak_matrix[:, i])
+                counter += 1
+                old_i = i
+            indices_to_merge.append((old_idx, len(conversionObject.joined.idx)))
+            
+            a = np.stack(added_indices)
+        else:
+            a = np.transpose(accessibility_peak_matrix, axes=[1, 0])
+        
+        a = a[:, None, :]
+
+        s = time.time()
+        
+        out = compute_casv(gen_to_list, a, radii)
+
+        casv_len = out.shape[1]
+        num_cells = out.shape[3]
+        num_regions = out.shape[0]
+        num_celltypes = out.shape[2]
+        num_targets = len(self.dataset.targets) if 'DNase' in self.dataset.targets else len(self.dataset.targets) + 1
+        total_targets = num_targets
+
+        s = time.time()
+
+        def first_substring(strings, substring, other, other2):
+#             return next(i for i, string in enumerate(strings) if substring in string)
+            for i, s in enumerate(strings):
+                if substring in s:
+                      return i
+            return None
+
+        for region in tqdm(range(num_regions)):
+            for cell in range(num_cells):
+
+                selected_gen = to_stack[cell, region, :]
+                naming_scheme = names[cell, region][0]
+                selected_casv = out[region, :, :, cell]
+
+                len_feats_per_celltype = int(selected_gen[0].shape[0] / num_celltypes) # 24 / 2 = 12
+
+                old_sg = selected_gen
+                
+
+                for celltype in range(num_celltypes):
+                    idx = (num_targets + 4) * celltype
+                    if idx >= len(selected_gen[0]):
+                        break
+                    num_targets = first_substring(naming_scheme[idx:idx+len_feats_per_celltype+total_targets], '_agree', naming_scheme[idx-10:idx+len_feats_per_celltype+10], (idx, len_feats_per_celltype))
+                    casv_cell = selected_casv[:, celltype]
+                    selected_gen[0][idx + num_targets : idx + num_targets + 4] = casv_cell
+        
+
+        results = []
```

Comment:
instead of returning a matrix, we can have a parameter that specifies where to write an h5 file. then we can iteratively write each slice to the h5 file